File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/AuthenticatedGreeterServer.scala
/*
 * Copyright (C) 2018-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package example.myapp.helloworld

import akka.actor.ActorSystem
import akka.grpc.GrpcServiceException
import akka.grpc.scaladsl.ServerInterceptor
import akka.grpc.scaladsl.ServerReflection
import akka.grpc.scaladsl.ServiceHandler
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse
import akka.http.scaladsl.server.Route
import akka.http.scaladsl.server.Directives._
import com.typesafe.config.ConfigFactory
import example.myapp.helloworld.grpc._
import io.grpc.Status

import scala.concurrent.{ ExecutionContext, Future }

object AuthenticatedGreeterServer {
  def main(args: Array[String]): Unit = {
    // Important: enable HTTP/2 in ActorSystem's config
    // We do it here programmatically, but you can also set it in the application.conf
    val conf =
      ConfigFactory.parseString("akka.http.server.enable-http2 = on").withFallback(ConfigFactory.defaultApplication())
    val system = ActorSystem("HelloWorld", conf)
    new AuthenticatedGreeterServer(system).run()
    // ActorSystem threads will keep the app alive until `system.terminate()` is called
  }
}

class AuthenticatedGreeterServer(system: ActorSystem) {
  def run(): Future[Http.ServerBinding] = {
    // Akka boot up code
    implicit val sys: ActorSystem = system
    implicit val ec: ExecutionContext = sys.dispatcher

    //#http-route
    // A Route to authenticate with
    val authenticationRoute: Route = path("login") {
      get {
        complete("Psst, please use token XYZ!")
      }
    }
    //#http-route

    //#interceptor
    // An interceptor that rejects calls without the right token
    val requireToken: ServerInterceptor = (serviceName, methodName, request, next) =>
      request.headers.find(_.name == "token") match {
        case Some(header) if header.value == "XYZ" => next(request)
        case _ =>
          Future.failed(
            new GrpcServiceException(
              Status.UNAUTHENTICATED.withDescription(s"Missing or invalid token for $serviceName/$methodName")))
      }
    //#interceptor

    //#grpc-protected
    // Create service handlers, the interceptor only applies to the greeter service
    val handler: HttpRequest => Future[HttpResponse] =
      ServiceHandler.concatOrNotFound(
        ServerInterceptor.intercept(GreeterServiceHandler.partial(new GreeterServiceImpl()), requireToken),
        ServerReflection.partial(List(GreeterService)))

    // As a Route
    val handlerRoute: Route = handle(handler)
    //#grpc-protected

    //#combined
    val route = concat(authenticationRoute, handlerRoute)

    // Bind service handler servers to localhost:8082
    val binding = Http().newServerAt("127.0.0.1", 8082).bind(route)
    //#combined

    // report successful binding
    binding.foreach { binding => println(s"gRPC server bound to: ${binding.localAddress}") }

    binding
  }
}

File: plugin-tester-java/src/main/java/example/myapp/helloworld/AuthenticatedGreeterServer.java
/*
 * Copyright (C) 2018-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package example.myapp.helloworld;

import java.util.Collections;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import akka.grpc.GrpcServiceException;
import akka.grpc.javadsl.ServerInterceptor;
import akka.grpc.javadsl.ServerInterceptors;
import akka.grpc.javadsl.ServerReflection;
import akka.grpc.javadsl.ServiceHandler;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;

import akka.actor.ActorSystem;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.http.javadsl.server.Route;
import akka.japi.function.Function;
import akka.stream.SystemMaterializer;
import akka.stream.Materializer;
import io.grpc.Status;

import example.myapp.helloworld.grpc.GreeterService;
import example.myapp.helloworld.grpc.GreeterServiceHandlerFactory;

import static akka.http.javadsl.server.Directives.*;

class AuthenticatedGreeterServer {
  public static void main(String[] args) throws Exception {
    // important to enable HTTP/2 in ActorSystem's config
    Config conf = ConfigFactory.parseString("akka.http.server.enable-http2 = on")
            .withFallback(ConfigFactory.defaultApplication());

    // Akka ActorSystem Boot
    ActorSystem sys = ActorSystem.create("HelloWorld", conf);

    run(sys).thenAccept(binding -> {
      System.out.println("gRPC server bound to: " + binding.localAddress());
    });

    // ActorSystem threads will keep the app alive until `system.terminate()` is called
  }

  public static CompletionStage<ServerBinding> run(ActorSystem sys) throws Exception {
    Materializer mat = SystemMaterializer.get(sys).materializer();

    //#http-route
    // A Route to authenticate with
    Route authentication = path("login", () ->
      get(() ->
        complete("Psst, please use token XYZ!")
      )
    );
    //#http-route

    //#interceptor
    // An interceptor that rejects calls without the right token
    ServerInterceptor requireToken = (serviceName, methodName, request, next) -> {
      boolean validToken = request.getHeader("token").map(header -> header.value().equals("XYZ")).orElse(false);
      if (validToken) {
        return next.apply(request);
      } else {
        CompletableFuture<HttpResponse> rejected = new CompletableFuture<>();
        rejected.completeExceptionally(new GrpcServiceException(
            Status.UNAUTHENTICATED.withDescription("Missing or invalid token for " + serviceName + "/" + methodName)));
        return rejected;
      }
    };
    //#interceptor

    //#grpc-protected
    // Create service handlers, the interceptor only applies to the greeter service
    GreeterService impl = new GreeterServiceImpl(mat);
    Function<HttpRequest, CompletionStage<HttpResponse>> handler =
      ServiceHandler.concatOrNotFound(
        ServerInterceptors.intercept(
          GreeterService.description,
          GreeterServiceHandlerFactory.create(impl, sys),
          sys,
          requireToken),
        ServerReflection.create(Collections.singletonList(GreeterService.description), sys));

    // As a Route
    Route handlerRoute = handle(handler);
    //#grpc-protected

    //#combined
    Route finalRoute = concat(
      authentication,
      handlerRoute
    );

    return Http.get(sys)
      .newServerAt("127.0.0.1", 8090)
      .bind(finalRoute);
    //#combined
  }
}
